baseplate/server/queue_consumer.py

```diff
@@ -0,0 +1,268 @@
+                # wait for all of our workers to finish before stopping the
+                # server.
+                message = self.work_queue.get(timeout=self._queue_timeout)
+            except queue.Empty:
+                pass
+            else:
+                # Ensure that if self.message_handler.handle throws a `queue.Empty`
+                # error, that bubbles up and is not treated as though `self.work_queue`
+                # is empty
+                self.message_handler.handle(message)
+        logger.debug("Consumer <%s> stopping.", self.id)
+
+
+class QueueConsumerServer:
+    """Server for running long-lived queue consumers."""
+
+    def __init__(
+        self, pump: PumpWorker, handlers: Sequence[QueueConsumer], healthcheck_server: StreamServer
+    ):
+        self.pump = pump
+        self.handlers = handlers
+        self.healthcheck_server = healthcheck_server
+        self.pump_thread = Thread(target=self.pump.run, daemon=True)
+        self.threads = [Thread(target=handler.run) for handler in self.handlers]
+        self.started = False
+        self.stopped = False
+        self._watch_period = 30
+
+    @classmethod
+    def new(
+        cls, max_concurrency: int, consumer_factory: QueueConsumerFactory, listener: socket.socket
+    ) -> "QueueConsumerServer":
+        """Build a new QueueConsumerServer."""
+        # We want to give some headroom on the queue so our handlers can grab
+        # a new message right after they finish so we keep an extra
+        # max_concurrency / 2 messages in the queue.
+        maxsize = max_concurrency + max_concurrency // 2
+        work_queue: queue.Queue = queue.Queue(maxsize=maxsize)
+        handlers = [
+            QueueConsumer(
+                work_queue=work_queue, message_handler=consumer_factory.build_message_handler()
+            )
+            for _ in range(max_concurrency)
+        ]
+        return cls(
+            pump=consumer_factory.build_pump_worker(work_queue),
+            handlers=handlers,
+            healthcheck_server=consumer_factory.build_health_checker(listener),
+        )
+
+    def _terminate(self) -> None:
+        """Send a SIGTERM signal to ourselves so baseplate can call `stop`."""
+        assert self.started
+        if not self.stopped:
+            os.kill(os.getpid(), signal.SIGTERM)
+
+    def _watch(self) -> None:
+        """Watches the threads to see if they have stopped prematurely.
+
+        Loops through the pump thread and handler threads periodically to check
+        if they are still alive and terminates the server if one is no longer
+        alive.
+        """
+        assert self.started
+        watched = [self.pump_thread] + self.threads
+        while not self.stopped:
+            for thread in watched:
+                if not thread.is_alive():
+                    logger.debug("Dead thread found, terminating server.")
+                    self._terminate()
```

Comment:
will we get any metrics around this? will we be able to see if the servers are constantly crashing (other than reading the logs)?

Reply:
AFAIK, we don't really have a good way to report metrics outside of the context of a request since the actual sending of metrics is tied to that lifecyle, so not specifically here.

1. We should have metrics on errors in the handler threads which would cause this
2. I would _like_ to support metrics like this if there's a way to @spladug ?

Reply:
oh, I missed this one. We've got the runtime monitor stuff now! You can report server-level metrics there. 

That said, if we just want to see if it's in a crash loop, the infra can tell us how long the process has been alive.